lib/datadog/tracing/sampling/span/rule.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'datadog/tracing/sampling/span/ext'
+
+module Datadog
+  module Tracing
+    module Sampling
+      module Span
+        # Span sampling rule that applies a sampling rate if the span
+        # matches the provided {Matcher}.
+        # Additionally, a rate limiter is also applied.
+        #
+        # If a span does not conform to the matcher, no changes are made.
+        class Rule
+          attr_reader :matcher, :sample_rate, :rate_limit
+
+          # Creates a new span sampling rule.
+          #
+          # @param [Sampling::Span::Matcher] matcher whether this rule applies to a specific span
+          # @param [Float] sample_rate span sampling ratio, between 0.0 (0%) and 1.0 (100%).
+          # @param [Numeric] rate_limit maximum number of spans sampled per second. Negative numbers mean unlimited spans.
+          def initialize(
+            matcher,
+            sample_rate: Span::Ext::DEFAULT_SAMPLE_RATE,
+            rate_limit: Span::Ext::DEFAULT_MAX_PER_SECOND
+          )
+
+            @matcher = matcher
+            @sample_rate = sample_rate
+            @rate_limit = rate_limit
+
+            @sampler = Sampling::RateSampler.new
+            @sampler.sample_rate = sample_rate
+            @rate_limiter = Sampling::TokenBucket.new(rate_limit)
+          end
+
+          # This method should only be invoked for spans that are part
+          # of a trace that has been dropped by trace-level sampling.
+          # Invoking it for other spans will cause incorrect sampling
+          # metrics to be reported by the Datadog App.
+          #
+          # Returns `true` if the provided span is sampled.
+          # If the span is dropped due to sampling rate or rate limiting,
+          # it returns `false`.
+          #
+          # Returns `nil` if the span did not meet the matching criteria by the
+          # provided matcher.
+          #
+          # This method modifies the `span` if it matches the provided matcher.
+          #
+          # @param [Datadog::Tracing::SpanOperation] span_op span to be sampled
+          # @return [Boolean] should this span be sampled?
+          # @return [nil] span did not satisfy the matcher, no changes are made to the span
+          def sample!(span_op)
+            return nil unless @matcher.match?(span_op)
+
+            if @sampler.sample?(span_op) && @rate_limiter.allow?(1)
+              span_op.set_metric(Span::Ext::TAG_MECHANISM, Span::Ext::MECHANISM_SPAN_SAMPLING_RATE)
+              span_op.set_metric(Span::Ext::TAG_RULE_RATE, @sample_rate)
+              span_op.set_metric(Span::Ext::TAG_MAX_PER_SECOND, @rate_limit)
+              true
+            else
+              false
+            end
```

Comment:
Minor: Is it possible to avoid the `nil`/`true`/`false` return and make this only a `true`/`false` result?

Having both `nil` and `false` floating around may be somewhat confusing, although I haven't seen the code for the caller, so this may actually be required... If we really need to treat them differently, it may be worth just using a bunch of symbols (something along the lines of `:sample`, `:dont_sample`, `:ignore`, for instance).

Reply:
I use the fact that is is `nil` here: https://github.com/DataDog/dd-trace-rb/pull/2098/files#diff-d196675ce6d3b5f69dd221b27626cbd863cb54ab43ec5ebdf23906aa6b7236beR38

> :sample, :dont_sample, :ignore

`true` `false` `nil` 🙂

I'd like to confirm, do you think that `:sample`, `:dont_sample`, `:ignore` are strongly better than `Bool|Nil`?

Reply:
> I'd like to confirm, do you think that :sample, :dont_sample, :ignore are strongly better than Bool|Nil?

In situations where `false` and `nil` have a semantic difference, I do! Mostly because while somewhere in the back of our minds we know there is a difference, we also know that most of the time we don't need to care, so I find that using the symbols screams a lot louder "THIS IS DIFFERENT AND YOU SHOULD HANDLE IT DIFFERENTLY" than even tests or comments may.

(There's a running joke on some old developer websites about [bool being defined as true/false/file not found](https://thedailywtf.com/articles/what_is_truth_0x3f_) and this is somewhat a similar situation)